pkg/exchange/coinbase/exchage.go

```diff
@@ -0,0 +1,241 @@
+		new_orders, err := e.client.GetOrders(ctx, localSymbol, status, paginationLimit, &sortedBy, &sorting, &before)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get orders while paginating")
+		}
+		if len(new_orders) < paginationLimit {
+			done = true
+		}
+		cbOrders = append(cbOrders, new_orders...)
+	}
+	return cbOrders, nil
+}
+
+func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) error {
+	failedOrderIDs := make([]string, 0)
+	for _, order := range orders {
+		req := e.client.NewCancelOrderRequest(order.UUID)
+		res, err := req.Do(ctx)
+		if err != nil {
+			log.WithError(err).Errorf("failed to cancel order: %v", order.UUID)
+			failedOrderIDs = append(failedOrderIDs, order.UUID)
+		}
+		log.Infof("order %v has been cancelled", res)
+	}
+	if len(failedOrderIDs) > 0 {
+		return errors.Errorf("failed to cancel orders: %v", failedOrderIDs)
+	}
+	return nil
+}
+
+// ExchangeMarketDataService
+func (e *Exchange) NewStream() types.Stream {
+	return nil
+}
+
+func (e *Exchange) QueryMarkets(ctx context.Context) (types.MarketMap, error) {
+	markets, err := e.client.GetMarketInfo(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get markets")
+	}
+	marketMap := make(types.MarketMap)
+	for _, m := range markets {
+		marketMap[toGlobalSymbol(m.ID)] = *toGlobalMarket(&m)
+	}
+	return marketMap, nil
+}
+
+func (e *Exchange) QueryTicker(ctx context.Context, symbol string) (*types.Ticker, error) {
+	return nil, nil
+}
+
+func (e *Exchange) QueryTickers(ctx context.Context, symbol ...string) (map[string]types.Ticker, error) {
+	return nil, nil
+}
+
+func (e *Exchange) QueryKLines(ctx context.Context, symbol string, interval types.Interval, options types.KLineQueryOptions) ([]types.KLine, error) {
+	return nil, nil
+}
+
+// ExchangeOrderQueryService
+func (e *Exchange) QueryOrder(ctx context.Context, q types.OrderQuery) (*types.Order, error) {
+	cbOrder, err := e.client.GetSingleOrder(ctx, q.OrderID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get order: %v", q.OrderID)
+	}
+	return toGlobalOrder(cbOrder), nil
+}
+
+func (e *Exchange) QueryOrderTrades(ctx context.Context, q types.OrderQuery) ([]types.Trade, error) {
+	cbTrades, err := e.queryOrderTradesByPagination(ctx, q.OrderID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get order trades: %v", q.OrderID)
+	}
+	trades := make([]types.Trade, 0, len(cbTrades))
+	for _, cbTrade := range cbTrades {
+		trades = append(trades, *toGlobalTrade(&cbTrade))
+	}
+	return trades, nil
+}
+
+func (e *Exchange) queryOrderTradesByPagination(ctx context.Context, orderID string) (api.TradeSnapshot, error) {
+	paginationLimit := 100
+	cbTrades, err := e.client.GetOrderTrades(ctx, orderID, paginationLimit, nil)
+	if err != nil {
+		return nil, err
+	}
+	if len(cbTrades) < paginationLimit {
+		return cbTrades, nil
+	}
+	done := false
+	if len(cbTrades) < paginationLimit || len(cbTrades) == 0 {
+		done = true
+	}
```

Comment:
Is it removable?

Reply:
In order to get all trades from Coinbase Exchange API, applying pagination is a must.
See:
- https://docs.cdp.coinbase.com/exchange/reference/exchangerestapi_getfills
- https://docs.cdp.coinbase.com/exchange/docs/rest-pagination